File: LeetCode/101-200/152. Maximum Product Subarray/src/Solution.java
class Solution {
    public int maxProduct(int[] nums) {
        int totatMax = nums[0];
        int currMax = nums[0];
        int currMin = nums[0];
        int temp;
        for (int i = 1; i < nums.length; i++) {
            temp = Math.max(nums[i], Math.max(nums[i] * currMax, nums[i] * currMin));
            currMin = Math.min(nums[i], Math.min(nums[i] * currMax, nums[i] * currMin));
            currMax = temp;
            totatMax = Math.max(totatMax, currMax);
        }
        return totatMax;
    }
}

File: LeetCode/101-200/152. Maximum Product Subarray/Solution.cpp
class Solution {
public:
    int maxProduct(vector<int>& nums) {
        int n = nums.size();
        int maxProd = nums[0];
        int maxPosPro = maxProd;
        int maxNegPro = maxProd;
        int temp;
        for (int i = 1; i < n; ++i) {
            temp = maxPosPro;
            maxPosPro = max(nums[i], max(maxPosPro * nums[i], maxNegPro * nums[i]));
            maxNegPro = min(nums[i], min(temp * nums[i], maxNegPro * nums[i]));
            maxProd = max(maxProd, maxPosPro);
        }
        return maxProd;
    }
};
